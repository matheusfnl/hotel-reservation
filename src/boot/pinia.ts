import { boot } from 'quasar/wrappers'
import { createPinia, PiniaVuePlugin } from 'pinia'

export default boot(({ app }) => {
  const pinia = createPinia()

  app.use(PiniaVuePlugin)
  app.use(pinia)
})
