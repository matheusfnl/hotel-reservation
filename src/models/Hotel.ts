import { createI18n } from 'vue-i18n'
import messages from '@/i18n'

import type { Amenities, Hotel as HotelInterface } from '@/types'
import { formatValue } from '@/helpers/number-to-value'

const i18n = createI18n({
  locale: 'pt-BR',
  legacy: false,
  messages,
})

export class Hotel {
  private id: number
  private favorite: boolean
  private name: string
  private description: string
  private stars: string
  private thumb: string
  private amenities: Amenities[] | null
  private hasBreakFast: boolean
  private hasRefundableRoom: boolean
  private hasAgreement: boolean
  private nonRefundable: boolean | null
  private address: {
    fullAddress: string
    city: string
    district: string
    street: string
    state: string
    zipCode: string | null
  }
  private images: string[]
  private deals: unknown
  private roomsQuantity: number
  private price: number

  constructor(data: HotelInterface) {
    this.id = data.id
    this.favorite = data.favorite
    this.name = data.name
    this.description = data.description
    this.stars = data.stars
    this.thumb = data.thumb
    this.amenities = data.amenities
    this.hasBreakFast = data.hasBreakFast
    this.hasRefundableRoom = data.hasRefundableRoom
    this.hasAgreement = data.hasAgreement
    this.nonRefundable = data.nonRefundable
    this.address = data.address
    this.images = data.images
    this.deals = data.deals
    this.roomsQuantity = data.roomsQuantity
    this.price = data.price
  }

  public get getId(): number {
    return this.id
  }

  public get getName(): string {
    return this.name
  }

  public get getDescription(): string {
    return this.description
  }

  public get getImages(): string[] {
    return this.images
  }

  public get getHotelLocation(): string {
    const { city, district, street } = this.address

    return `${city}, ${district}, ${street}.`
  }

  public get getFullHotelLocation(): string {
    const { fullAddress, city, state, zipCode } = this.address

    return `${fullAddress}, ${city} - ${state}${zipCode ? `, ${zipCode}` : ''}`
  }

  public get getRating(): string {
    return this.stars
  }

  public get getAmenities(): Amenities[] {
    let amenities = this.amenities ? [...this.amenities] : []
    amenities = amenities.filter((amenity) => !Array.isArray(amenity))

    if (this.hasBreakFast && !amenities.find((amenity) => amenity.key === 'BREAKFAST')) {
      amenities.push({
        key: 'BREAKFAST',
        label: 'Café da manhã',
      })
    }

    return amenities || []
  }

  public get isRefundable(): boolean {
    return this.hasRefundableRoom && !this.nonRefundable
  }

  public get getFullPrice(): string {
    return formatValue(this.price)
  }

  public get getPerNightText(): string {
    const value = this.price / 3

    return `${i18n.global.t('hotel.hotel.currency')}${formatValue(value)}${i18n.global.t('hotel.hotel.per_night')}`
  }
}
