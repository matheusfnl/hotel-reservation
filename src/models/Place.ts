import type { Places, PlaceState } from '@/types'

export class Place {
  readonly name: string
  readonly state: PlaceState
  readonly placeId: number

  constructor(data: Places) {
    this.name = data.name
    this.state = data.state
    this.placeId = data.placeId
  }
}
