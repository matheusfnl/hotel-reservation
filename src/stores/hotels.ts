import { defineStore } from 'pinia'

import type { HotelsByPlace, Hotel } from '@/types'

import hotelsMock from '@/mock/hotel.json'

interface FetchHotelsParams {
  placeId?: number | string
  page?: number
  name?: string
  sort?: string
}

const allPlaceHotels = [...hotelsMock] as HotelsByPlace[]

export const useHotelsStore = defineStore('hotels', {
  state: () => ({ hotels: [] as Hotel[] }),
  actions: {
    async fetchHotels({ placeId, page = 1, name, sort }: FetchHotelsParams = {}) {
      try {
        const filterHotels = (): Hotel[] => {
          let placeHotels = placeId
            ? allPlaceHotels.find((hotel) => hotel.placeId === +placeId)?.hotels
            : allPlaceHotels.flatMap((place) => place.hotels)

          if (!placeHotels) {
            return []
          }

          if (name) {
            placeHotels = placeHotels.filter((hotel) =>
              hotel.name.toLowerCase().includes(name.toLowerCase()),
            )
          }

          if (['high_price', 'low_price'].includes(sort as string)) {
            return sort === 'low_price'
              ? placeHotels?.sort((a, b) => +a.price - +b.price)
              : placeHotels?.sort((a, b) => +b.price - +a.price)
          }

          if (sort === 'best_rated') {
            return placeHotels?.sort((a, b) => +b.stars - +a.stars)
          }

          return placeHotels?.sort((a, b) =>
            a.favorite !== b.favorite ? +b.favorite - +a.favorite : +b.stars - +a.stars,
          )
        }

        await new Promise((resolve) => setTimeout(resolve, 2000))

        const placeHotels = filterHotels()
        const itemsIndex = (page - 1) * 10

        this.hotels = placeHotels ? placeHotels.slice(itemsIndex, itemsIndex + 10) : []
      } catch (error: unknown) {
        return error
      }
    },
  },
})
