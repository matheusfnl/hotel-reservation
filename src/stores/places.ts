import { defineStore } from 'pinia'

import { Place } from '@/models/Place'
import type { Places } from '@/types'

import placesMock from '@/mock/place.json'

const places = placesMock as Places[]

export const usePlacesStore = defineStore('places', {
  state: () => ({ places: [] as Place[] }),
  actions: {
    async fetchPlaces() {
      try {
        await new Promise((resolve) => setTimeout(resolve, 2000))

        this.places = places.map((place) => new Place(place))
      } catch (error: unknown) {
        return error
      }
    },
  },
})
