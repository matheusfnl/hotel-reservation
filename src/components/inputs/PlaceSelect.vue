<template>
  <q-select
    v-model="placeSearch"
    use-input
    hide-selected
    fill-input
    outlined
    :dense="dense"
    clearable
    hide-dropdown-icon
    input-debounce="0"
    :placeholder="$t('hotel.search.inputs.destination.placeholder')"
    :options="filteredPlaces"
    option-label="formattedLabel"
    @filter="filterPlaces"
    @update:model-value="$emit('update:modelValue', $event)"
  >
    <template v-slot:no-option>
      <q-item>
        <q-item-section class="text-grey">
          {{ $t('hotel.search.inputs.destination.empty') }}
        </q-item-section>
      </q-item>
    </template>
  </q-select>
</template>

<script lang="ts" setup>
import { ref, onMounted } from 'vue'
import { useRoute } from 'vue-router'

import type { FormattedPlaces } from '@/types'

import { usePlacesStore } from '@/stores/places'

const placeStore = usePlacesStore()

const route = useRoute()

const emit = defineEmits<{
  'update:modelValue': [value: FormattedPlaces | null]
}>()

defineProps({
  dense: {
    type: Boolean,
    default: true,
  },
})

const placeSearch = ref<FormattedPlaces | null>(null)
const filteredPlaces = ref<FormattedPlaces[]>([])

const filterPlaces = (val: string, update: (fn: () => void) => void) => {
  const formattedPlaces = placeStore.places.map((place) => ({
    formattedLabel: `${place.name}, ${place.state.shortname}`,
    ...place,
  }))

  update(() => {
    const search = val.toLowerCase()
    filteredPlaces.value = formattedPlaces.filter(
      (place) =>
        place.formattedLabel.toLowerCase().includes(search) ||
        place.state.name.toLowerCase().includes(search),
    )
  })
}

const setPlace = (placeId: string | number) => {
  const formattedPlaces = placeStore.places.map((place) => ({
    formattedLabel: `${place.name}, ${place.state.shortname}`,
    ...place,
  }))

  emit('update:modelValue', formattedPlaces.find((place) => +place.placeId === +placeId) || null)
}

onMounted(async () => {
  await placeStore.fetchPlaces()

  if (route.params.placeId) {
    setPlace(route.params.placeId as string)
  }
})

defineExpose({
  setPlace,
})
</script>
