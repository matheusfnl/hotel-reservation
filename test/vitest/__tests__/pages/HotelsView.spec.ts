import { describe, it, expect, beforeEach, vi } from 'vitest'
import { setActivePinia, createPinia } from 'pinia'
import { mount } from '@vue/test-utils'
import { QBtn } from 'quasar'
import { createRouter, createWebHistory } from 'vue-router'

import { useHotelsStore } from '@/stores/hotels'
import HotelsView from '@/pages/HotelsView.vue'
import PlaceSelect from '@/components/inputs/PlaceSelect.vue'

import type { Hotel as HotelInterface } from '@/types'
import { Hotel } from '@/models/Hotel'

const mockHotelData: HotelInterface = {
  id: 1,
  favorite: false,
  name: 'Hotel Teste',
  description: 'Descrição do hotel',
  stars: '4.5',
  thumb: 'thumb.jpg',
  amenities: [{ key: 'ROOM_SERVICE', label: 'Wi-Fi' }],
  hasBreakFast: true,
  hasRefundableRoom: true,
  hasAgreement: false,
  nonRefundable: null,
  address: {
    fullAddress: 'Rua Exemplo, 123',
    city: 'São Paulo',
    district: 'Centro',
    street: 'Rua Exemplo',
    state: 'SP',
    zipCode: '12345-678',
  },
  images: ['image1.jpg', 'image2.jpg'],
  deals: null,
  roomsQuantity: 10,
  price: 1000,
}

const mockHotel = new Hotel(mockHotelData)

const router = createRouter({
  history: createWebHistory(),
  routes: [],
})

describe('HotelsView', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
    const hotelsSore = useHotelsStore()
    hotelsSore.hotels = [mockHotel]

    vi.spyOn(router, 'push').mockImplementation(() => Promise.resolve())
  })

  const mountWrapper = () => {
    return mount(HotelsView, {
      global: {
        plugins: [router],
      },
    })
  }

  it('should render correctly', () => {
    const wrapper = mountWrapper()
    const html = wrapper
      .html()
      .replace(/id="f_[^"]+"/g, 'id="static-id"')
      .replace(/for="f_[^"]+"/g, 'for="static-for"')
      .replace(/aria-controls="f_[^"]+"/g, 'aria-controls="static-aria-controls"')

    expect(html).toMatchSnapshot()
  })

  it('should update place and search on button click', async () => {
    const wrapper = mountWrapper()
    const selectComponent = wrapper.findComponent(PlaceSelect)
    const select = selectComponent.find('input')
    const button = wrapper.findComponent(QBtn)

    await select.setValue('São Paulo')
    await select.trigger('input')
    await button.trigger('click')

    // eslint-disable-next-line @typescript-eslint/unbound-method
    expect(router.push).toHaveBeenCalledWith({
      name: 'hotels.place',
      params: { placeId: undefined },
      query: {},
    })
  })
})
