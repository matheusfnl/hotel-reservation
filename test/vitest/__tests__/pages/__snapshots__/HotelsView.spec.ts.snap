// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelsView > should render correctly 1`] = `
"<div data-v-7ea6272d="" class="q-layout q-layout--standard" style="min-height: 768px;" tabindex="-1">
  <!---->
  <!---->
  <form data-v-7ea6272d="" novalidate="">
    <div data-v-7ea6272d="" class="q-card bg-white rounded-borders-md">
      <div class="q-card__section q-card__section--vert q-px-md q-py-sm">
        <h2 class="text-h5 q-ma-none">Reservar hotel</h2>
      </div>
      <hr class="q-separator q-separator--horizontal" aria-orientation="horizontal">
      <div class="q-card__section q-card__section--vert column q-gutter-md">
        <div>
          <div data-v-7ea6272d="" class="row q-col-gutter-x-md"><label data-v-7ea6272d="" class="col-12 col-sm-6"><span data-v-7ea6272d="" class="text-weight-medium text-grey-7">Destino <span data-v-7ea6272d="" class="text-red">*</span></span><label data-v-cf369733="" data-v-7ea6272d="" class="q-field row no-wrap items-start q-field--outlined q-select q-field--auto-height q-select--with-input q-select--without-chips q-select--single q-field--dense place-select" for="static-for">
                <!---->
                <div class="q-field__inner relative-position col self-stretch">
                  <div class="q-field__control relative-position row no-wrap" tabindex="-1">
                    <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip">
                      <div class="q-field__native row items-center"><input class="q-field__input q-placeholder col" type="search" tabindex="0" role="combobox" aria-readonly="false" aria-autocomplete="list" aria-expanded="false" aria-controls="static-aria-controls" placeholder="Cidade, UF" selected-item-label="selectedLabel" id="static-id" value=""></div>
                      <!---->
                    </div>
                    <!---->
                    <!---->
                  </div>
                  <!---->
                </div>
                <!---->
              </label>
              <!--v-if-->
            </label><label data-v-7ea6272d="" class="col-12 col-sm-6"><span data-v-7ea6272d="" class="text-weight-medium text-grey-7">Nome</span><label data-v-7ea6272d="" class="q-field row no-wrap items-start q-field--outlined q-input q-field--dense" for="static-for">
                <!---->
                <div class="q-field__inner relative-position col self-stretch">
                  <div class="q-field__control relative-position row no-wrap" tabindex="-1">
                    <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input class="q-field__native q-placeholder" tabindex="0" placeholder="Hotel..." id="static-id" type="text" value="">
                      <!---->
                    </div>
                  </div>
                  <!---->
                </div>
                <!---->
              </label></label></div>
        </div>
        <div class="row justify-end"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle q-btn--rounded bg-primary text-white q-btn--actionable q-focusable q-hoverable q-btn--no-uppercase col-12 col-sm-auto" style="font-size: 14px; min-width: 180px;" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row">Buscar </span>
            <transition-stub name="q-transition--fade" appear="false" persisted="false" css="true">
              <!---->
            </transition-stub>
          </button></div>
      </div>
    </div>
  </form>
  <div data-v-7ea6272d="" class="row items-center justify-between q-py-md" style="row-gap: 16px;">
    <div data-v-7ea6272d="" class="q-breadcrumbs text-brown">
      <div class="flex items-center justify-start q-gutter-sm">
        <div class="flex items-center text-primary"><span class="q-breadcrumbs__el q-link flex inline items-center relative-position q-link--focusable">Início</span></div>
        <div class="q-breadcrumbs__separator"><i class="q-icon text-primary notranslate material-icons" style="font-size: 1.5em;" aria-hidden="true" role="presentation">chevron_right</i></div>
        <div class="flex items-center text-primary"><span class="q-breadcrumbs__el q-link flex inline items-center relative-position q-link--focusable">Hotéis</span></div>
        <div class="q-breadcrumbs__separator"><i class="q-icon text-primary notranslate material-icons" style="font-size: 1.5em;" aria-hidden="true" role="presentation">chevron_right</i></div>
        <div class="flex items-center q-breadcrumbs--last"><span class="q-breadcrumbs__el q-link flex inline items-center relative-position q-link--focusable">Hospedagem</span></div>
      </div>
    </div><span data-v-7ea6272d="">Organizar por <span class="text-primary text-bold text-italic cursor-pointer">Recomendados <i class="q-icon notranslate material-icons q-mr-sm" aria-hidden="true" role="presentation">expand_more</i><!----></span></span>
  </div>
  <div data-v-7ea6272d="" class="row q-col-gutter-y-md">
    <div data-v-7ea6272d="" class="col-12">
      <div data-v-be666f43="" data-v-7ea6272d="" class="q-card bg-white rounded-borders-md row">
        <div data-v-be666f43="" class="q-card__section q-card__section--vert image-section q-pa-none col-12 col-sm-3 overflow-hidden">
          <div data-v-be666f43="" class="q-carousel q-panel-parent q-carousel--without-padding q-carousel--arrows-horizontal rounded-borders-md no-scroll" style="height: 256px;">
            <div class="q-carousel__slides-container">
              <transition-stub appear="false" persisted="false" css="true">
                <div class="q-panel scroll" style="--q-transition-duration: 300ms;" role="tabpanel">
                  <div data-v-be666f43="" class="q-carousel__slide q-pa-none" style="height: 256px;">
                    <div data-v-be666f43="" class="q-img q-img--no-menu full-width" style="height: 256px;" role="img">
                      <div style="padding-bottom: 56.24929688378895%;"></div>
                      <div class="q-img__container absolute-full"><img class="q-img__image q-img__image--with-transition q-img__image--current" style="object-fit: cover; object-position: 50% 50%;" height="256px" loading="lazy" fetchpriority="auto" aria-hidden="true" draggable="false" src="image1.jpg"></div>
                      <transition-stub name="q-transition--fade" appear="false" persisted="false" css="true">
                        <div class="q-img__loading absolute-full flex flex-center"><svg class="q-spinner text-primary q-spinner-mat" width="1em" height="1em" viewBox="25 25 50 50">
                            <circle class="path" cx="50" cy="50" r="20" fill="none" stroke="currentColor" stroke-width="5" stroke-miterlimit="10"></circle>
                          </svg></div>
                      </transition-stub>
                    </div>
                  </div>
                </div>
              </transition-stub>
            </div>
            <div class="q-carousel__control q-carousel__arrow q-carousel__prev-arrow q-carousel__prev-arrow--horizontal absolute flex flex-center"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-grey q-btn--actionable q-focusable q-hoverable q-btn--dense" tabindex="0" type="button" regular="true"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">chevron_left</i></span></button></div>
            <div class="q-carousel__control q-carousel__arrow q-carousel__next-arrow q-carousel__next-arrow--horizontal absolute flex flex-center"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-grey q-btn--actionable q-focusable q-hoverable q-btn--dense" tabindex="0" type="button" regular="true"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">chevron_right</i></span></button></div>
          </div>
        </div>
        <div data-v-be666f43="" class="q-card__section q-card__section--vert col-12 col-sm-5 col-md-6">
          <p data-v-be666f43="" class="text-h5 q-mb-none cursor-pointer clickable-title two-lines-truncate">Hotel Teste</p>
          <p data-v-be666f43="" class="text-grey-6">São Paulo, Centro, Rua Exemplo.</p>
          <div data-v-be666f43="" class="q-ma-none q-mb-sm row items-center q-gutter-x-sm">0 <div data-v-be666f43="" class="q-rating row inline items-center q-rating--non-editable text-yellow" style="font-size: 18px;" role="radiogroup" aria-readonly="true">
              <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 1"><i class="q-icon notranslate material-icons q-rating__icon text-yellow" aria-hidden="true" role="presentation">grade</i></div>
              <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 2"><i class="q-icon notranslate material-icons q-rating__icon text-yellow" aria-hidden="true" role="presentation">grade</i></div>
              <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 3"><i class="q-icon notranslate material-icons q-rating__icon text-yellow" aria-hidden="true" role="presentation">grade</i></div>
              <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 4"><i class="q-icon notranslate material-icons q-rating__icon text-yellow" aria-hidden="true" role="presentation">grade</i></div>
              <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 5"><i class="q-icon notranslate material-icons q-rating__icon text-yellow" aria-hidden="true" role="presentation">grade</i></div>
            </div>
            <hr data-v-be666f43="" class="q-separator q-separator--vertical bg-grey" aria-orientation="vertical"><i data-v-be666f43="" class="q-icon notranslate material-icons" style="font-size: 18px;" aria-hidden="true" role="presentation">room_service
              <!---->
            </i><i data-v-be666f43="" class="q-icon notranslate material-icons" style="font-size: 18px;" aria-hidden="true" role="presentation">bakery_dining
              <!---->
            </i>
          </div>
          <div data-v-be666f43="" class="q-chip row inline no-wrap items-center bg-grey text-white q-chip--colored q-chip--square text-weight-medium">
            <div class="q-chip__content col row no-wrap items-center q-anchor--skip">Reembolsável</div>
          </div>
        </div>
        <div data-v-be666f43="" class="q-card__section q-card__section--vert col-12 col-sm-4 col-md-3 relative-position">
          <hr data-v-be666f43="" class="q-separator q-separator--vertical bg-grey-3 position-left" style="margin-right: 8px; margin-left: 8px;" aria-orientation="vertical">
          <p data-v-be666f43="" class="q-mb-xs text-grey-6 text-caption">A partir de:</p>
          <p data-v-be666f43="" class="q-mb-xs">R$<span data-v-be666f43="" class="text-h5">1.000,00</span></p>
          <p data-v-be666f43="" class="q-mb-none text-grey-6 text-caption">R$333,33/noite</p>
          <p data-v-be666f43="" class="text-grey-7 q-mb-xl text-caption">Impostos inclusos</p>
          <siv data-v-be666f43="" class="row"><button data-v-be666f43="" class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle q-btn--rounded bg-primary text-white q-btn--actionable q-focusable q-hoverable q-btn--no-uppercase col-12 col-sm-auto" style="font-size: 14px; min-width: 140px;" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row">Selecionar</span></button></siv>
        </div>
      </div>
    </div><span data-v-7ea6272d="" class="full-width flex justify-center text-grey-6 empty-feedback q-ma-none">Todos os hotéis foram listados</span>
    <!--v-if-->
  </div>
</div>"
`;
